File: Data Structures & Algorithms/invert-a-binary-tree/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    // BFS solution, O(n) time and space
    TreeNode* invertTree(TreeNode* root) {
        if (!root) {
            return nullptr;
        }
        queue<TreeNode*> q;
        q.push(root);
        while (!q.empty()) {
            TreeNode* curr_node = q.front();
            q.pop();
            swap(curr_node->left, curr_node->right);
            if (curr_node->left) {
                q.push(curr_node->left);
            }
            if (curr_node->right) {
                q.push(curr_node->right);
            }
        }

        return root;
    }
};

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-4.py
class Solution:
    # one-pass binary search solution, O(log n) time, O(1) space
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1
        while l <= r:
            m = l + (r - l) // 2
            if nums[m] == target:
                return m
            
            # if we are at the left sorted portion
            if nums[m] >= nums[l]:
                # do regular binary search inside this portion
                if nums[l] <= target < nums[m]:
                    r = m - 1
                else:
                    l = m + 1

            # if we are at the right sorted portion
            else:
                # do regular binary search inside this portion
                if nums[m] < target <= nums[r]:
                    l = m + 1
                else:
                    r = m - 1

        return -1
